_posts/2015-03-stitch-fix-heart-unix.md

```diff
@@ -0,0 +1,670 @@
+
+## Files
+
+<a name="check-image-size"></a>
+### Check Image Size
+&#8212; [Dave Copeland](https://twitter.com/davetron5000/) *(OS X / bash)*
+
+#### Problem
+Check the sizes of our product images the pickers use to check what they are picking in our warehouse
+
+#### Solution
+Created a script to run curl on the argument to download the image, use ImageMagick to get its size and print that out in CSV.  Piped my input CSV of images into `xargs -n1 -P8 ./my_script.rb` to basically run my script 8-way parallel to get the job done as fast as I could without setting my machine on fire.
+
+##### Input
+```csv
+1,https://cdn.example.com/image_A.jpg
+5,https://cdn.example.com/image_B.jpg
+10,https://cdn.example.com/image_C.jpg
+```
+
+##### Command
+```
+cat my_input.csv | xargs -n1 -P8 ./check.rb > images_with_possible_issues.csv
+```
+
+ - `xargs` runs a command, feeding `xargs` `STDIN` as arguments to that command
+   - `echo ""foo"" | xargs ls` is the same as `""ls foo""`.  
+ - `-n1` says to run the given command once for each line of input.  
+  - Normally `xargs` will run many lines at once, so if you had a file with 10 rows in it called `""blah.csv""` and do `cat blah.csv | xargs curl`, it would likely run curl once with all 10 rows of blah.csv given to curl.  `-n1` means to just do them one at a time (so 10 runs of curl in this example).
+ - `-P8` says to parallelize it 8 ways.  
+
+So, I've got 8 instances of my script running at once.  Obviously, there are diminishing returns on parallelism, but since curl'ing images is mostly I/O bound this worked pretty well without compromising my machine.
+
+<a name="find-recently-changed-files"></a>
+### Find Recently Changed Files
+&#8212; [Greg Novak](https://www.linkedin.com/in/gsnovak/) *(OS X / bash)*
+
+#### Problem
+Some file changed but I don't know where.  Sometimes this is "I don't know where my web browser saved my file."  I want a list of the most recently changed file in a tree.
+
+#### Solution
+Find command as answered on [Stack Overflow](http://stackoverflow.com/questions/4561895/how-to-recursively-find-the-latest-modified-file-in-a-directory).
+
+##### Input
+`cd` to some directory
+
+##### Command
+```
+# OS X find:
+find . -type f -print0 | xargs -0 stat -f ""%m %N"" | sort -rn | head -1 | cut -f2- -d" "
+
+# GNU find:
+find . -type f -printf '%T@ %p\n' | sort -n | tail -1 | cut -f2- -d" "
+```
+
+ - The gnu `find` one does the heavy lifting inside the find command itself.  
+ - OS X `find` is a bit dumber so it just provides a list of files, gets modification times by calling stat on each one via `xargs`, then sort and clip.
+
+<a name="count-data-dump-number-of-lines"></a>
+### Count Data Dump Number Of Lines
+&#8212; [Deep Ganguli](https://twitter.com/dgangul1/) *(OS X / bash)*
+
+#### Problem
+How many lines of text are there in a data dump? This happens all the time, and I hate opening the data in a text editor and scrolling to the bottom.
```

Comment:
Replace these two sentences with:

A common question I have is: how many rows of data are in this file? The egregiously lazy method of obtaining an answer is to open a text editor, scroll to the bottom, and read off the last line number. This is inefficient.